frontend/src/Pages/Favorites.tsx

```diff
@@ -0,0 +1,150 @@
+import { useEffect, useState } from 'react';
+import { useAuth } from '@/useAuth';
+import API from '@/api/api';
+import FavoriteCard from '@/Components/FavoriteCard';
+import { CombinedFavorite } from '@/common';
+import Loading from '@/Components/Loading';
+import { mutate } from 'swr';
+
+export default function FavoritesPage() {
+    const { user } = useAuth();
+    const [favorites, setFavorites] = useState<CombinedFavorite[]>([]);
+    const [isLoading, setIsLoading] = useState<boolean>(true);
+    const [error, setError] = useState<string | null>(null);
+
+    useEffect(() => {
+        const fetchFavorites = async (): Promise<void> => {
+            try {
+                setIsLoading(true);
+                setError(null);
+                const response = await API.get('open-content/favorites');
+                interface ApiResponse {
+                    data: {
+                        id: number;
+                        content_id: number;
+                        name: string;
+                        type: 'library' | 'video';
+                        thumbnail_url: string;
+                        description: string;
+                        content_url: string;
+                        visibility_status: boolean;
+                        open_content_provider_id: number;
+                        channel_title?: string;
+                        provider_name?: string;
+                    }[];
+                }
+                const typedResponse = response as ApiResponse;
+                const mappedFavorites: CombinedFavorite[] =
+                    typedResponse.data.map((item) => ({
+                        id: item.id,
+                        content_id: item.content_id,
+                        name: item.name,
+                        type: item.type,
+                        thumbnail_url: item.thumbnail_url,
+                        description: item.description,
+                        visibility_status: item.visibility_status,
+                        open_content_provider_id: item.open_content_provider_id,
+                        content_url: item.content_url,
+                        channel_title: item.channel_title ?? '',
+                        provider_name: item.provider_name ?? ''
+                    }));
+                setFavorites(mappedFavorites);
+            } catch (err) {
+                setError('Error fetching favorites.');
+                console.error(err);
+            } finally {
+                setIsLoading(false);
+            }
+        };
+        void fetchFavorites();
+    }, [user]);
+
+    const handleUnfavorite = async (
+        contentId: number,
+        type: 'video' | 'library'
+    ) => {
+        try {
```

Comment:
line 66) calls from the API class don't need to be wrapped in try/catch block :)

(line 62) `type` is a keyword in TS, so we should avoid using it as a variable name.

lines 68-74 could be written more concisely as:

```ts
const endpoint = favType === 'video' ? `videos/${contentId}/favorite` : `open-content/${contentId}/save`;
```

Reply:
Looking at this code, you could actually probably do this unfavoriting within the favorite card component itself rather than in this higher level element. That would eliminate the lines of code of you checking to see if the favorite is in the favorites array since you can directly access that particular favorites information from that card. This, paired with the implementation of useSWR, would significantly reduce the amount of code that is in the `Favorites.tsx` file to around 30 total lines of code. Hard to explain without actually showing the code: 

Inside of `Favorites.tsx`:
```
export default function FavoritesPage() {
    const {
        data: favorites,
        isLoading,
        error,
        mutate
    } = useSWR<ServerResponseMany<CombinedFavorite>, AxiosError>(
        `/api/open-content/favorites`
    );

    if (isLoading) {
        return <Loading />;
    }

    return (
        <div className="px-8 pb-4">
            <h1>My Favorites</h1>
            {error || favorites == undefined ? (
                <div className="text-error">Error loading favorites</div>
            ) : favorites.data.length === 0 ? (
                <p>You have no favorites yet.</p>
            ) : (
                <div className="grid grid-cols-4 gap-6 mt-6">
                    {favorites.data.map((favorite) => (
                        <FavoriteCard
                            key={favorite.content_id}
                            favorite={favorite}
                            mutate={mutate}
                        />
                    ))}
                </div>
            )}
        </div>
    );
```

Inside of `FavoriteCard.tsx` 
```
interface FavoriteCardProps {
    favorite: CombinedFavorite;
    mutate: KeyedMutator<ServerResponseMany<CombinedFavorite>>;
}

const FavoriteCard: React.FC<FavoriteCardProps> = ({ favorite, mutate }) => {
    const navigate = useNavigate();
    const { toaster } = useToast();

    const handleUnfavorite = async () => {
        const { content_id, name, open_content_provider_id } = favorite;
        const endpoint =
            favorite.type === 'video'
                ? `videos/${content_id}/favorite`
                : `open-content/${content_id}/save`;

        const payload = {
            name,
            content_id,
            open_content_provider_id,
            content_url: `/api/proxy/libraries/${content_id}/`
        };

        const response = await API.put(endpoint, payload);

        if (response.success) {
            await mutate();
        } else {
            toaster('Failed to unfavorite item.', ToastState.error);
        }
    };

// REST OF CODE
```

follow up: do you even need to pass a payload into this api put? In libraries i just call the put and it should do the toggling for me in the backend. might be different- but check out `LibraryCard.tsx`

Reply:
Also, not sure if this was implemented the way it was because of the nested favoriting- and in that case let me know. Since i am only looking at the code that is in this branch, I'm not sure what you need/don't need in the case of favoriting nested pages.